# Inheritance-of-HEB-in-wheat

## Background:

## Data availability:

All RNA-seq data were uploaded to [Sequence Read Archive (SRA)](https://www.ncbi.nlm.nih.gov/sra) under the BioProject ID PRJNA1128551.

The wheat genome reference used to map/pseudoalign the RNA-seq data to is The IWGSC RefSeq v1.0 assembly available [here](https://urgi.versailles.inra.fr/download/iwgsc/IWGSC_RefSeq_Assemblies/v1.0/). 

The wheat genome annotation used to map SNPs to is The IWGSC RefSeq v1.1 annotation available [here](https://urgi.versailles.inra.fr/download/iwgsc/IWGSC_RefSeq_Annotations/v1.1/).

The triad identification file 

## Methodology:

Scripts are organised by Methods sections as outlined in the manuscript in the [scripts](scripts/) folder.


  ### 1. [HEB analysis](scripts/HEB/)


  ### 2. [Gene ontology term enrichment](scripts/Gene ontology term enrichment)


  ### 3. [Variant calling, genotyping and association between genotype and homoeolog expression](scripts/Variant calling, genotyping and association between genotype and homoeolog expression)

  
  ### 4. [SNP genotyping of F<sub>5</sub> lines for eQTL mapping](scripts/SNP genotyping and Detection of eQTL)


  ### 5. [Quantification of gene expression levels and correction for batch effects](scripts/Quantification of gene expression levels and correction for batch effects)


  ### 6. [Detection of eQTL](scripts/SNP genotyping and Detection of eQTL)

_Note: Sections 4 and 6 are coupled together due to overlapping code sections._

## Figures:

The code for recreating figure panels along with ready to plot data is organised by figure numbers as stated in the manuscript, in the [figures](figures/) folder.
